Contracts/Presale/PresaleContract.sol

```diff
@@ -23,98 +23,112 @@ contract PresaleContract is WithdrawContract {
+        mainWallet = owner;
+        amountBoughtInOneWei.push(FIRST_PRESALE_PRICE);
+        amountBoughtInOneWei.push(SECOND_PRESALE_PRICE);
+        amountBoughtInOneWei.push(THIRD_PRESALE_PRICE);
         
-        amounts.push(FIRST_PRESALE_AMOUNT * MINIMUM_TOKEN_VALUE_MULTIPLIER);
-        amounts.push(SECOND_PRESALE_AMOUNT * MINIMUM_TOKEN_VALUE_MULTIPLIER);
-        amounts.push(THIRD_PRESALE_AMOUNT * MINIMUM_TOKEN_VALUE_MULTIPLIER);
+        amountOfTokensLeft.push(FIRST_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
+        amountOfTokensLeft.push(SECOND_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
+        amountOfTokensLeft.push(THIRD_PRESALE_AMOUNT * EXTRA_DIGITS_MULTIPLIER);
     }
     
     mapping (address => Balance) usersBalance;
     Transaction[] transactions;
     
-    function buyTokens() public payable {
+    function () external payable {
         require(msg.value >= MINIMUM_PAYMENT_VALUE);
-        uint deposit = msg.value;
+        uint totalDeposit = msg.value;
         
-        uint8 index = 0;
-        uint profit = 0;
+        uint8 stage = 0;
+        uint currentInvestment = 0;
         
-        uint totalTokens = 0;
+        uint totalPurchasedTokens = 0;
         
-        while (deposit > 0 && index < 3) {
-            uint ethers = 0;
-            uint maxAmount = deposit * MINIMUM_TOKEN_VALUE_MULTIPLIER * prices[index] / 1 ether;
-            if (amounts[index] >= maxAmount) {
-                amounts[index] -= maxAmount;
-                ethers = deposit;
-                deposit = 0;
+        while (totalDeposit > 0 && stage < 3) {
+            uint currentStageDeposit = 0;
+            
+            uint maxTokensAmountForCurrentStage = totalDeposit * amountBoughtInOneWei[stage];
+            //checking free tokens amount for current stage
+            if (amountOfTokensLeft[stage] >= maxTokensAmountForCurrentStage) {
+                amountOfTokensLeft[stage] -= maxTokensAmountForCurrentStage;
+                currentStageDeposit = totalDeposit;
+                totalDeposit = 0;
             } else {
-                maxAmount = amounts[index];
-                amounts[index] = 0;
-                ethers = maxAmount * 1 ether / MINIMUM_TOKEN_VALUE_MULTIPLIER / prices[index];
-                deposit -= ethers;
+                maxTokensAmountForCurrentStage = amountOfTokensLeft[stage];
+                currentStageDeposit = maxTokensAmountForCurrentStage / amountBoughtInOneWei[stage];
                 
+                //if we have less tokens then min price we sell them for 1 wei
+                //example we have 101000 / 10^18 tokens 
+                //current price is 2500 tokens for 1 ether => 2500 / 10^18 tokens for 1 wei
+                //(101000 / 10^18 % (2500 / 10^18)) => (101000 % 2500) => 1000
+                //1000 > 0 so we sell 1000 / 10^18 tokens for 1 wei (minimum amount of ethers)
+                if ((maxTokensAmountForCurrentStage % amountBoughtInOneWei[stage]) > 0) {

```

Comment:
@Yaschenko this is nice, but i dont think this can happen. since we define maxTokensAmountForCurrentStage  as:
totalDeposit * amountBoughtInOneWei[stage];

so I think that it will always return 0 when modulo with amountBoughtInOneWei[stage];